### DISCUSSION TITLE
Implementing new pmd rule

### DISCUSSION THREAD

**Discussion by Prapti08** [pmd/pmd#4229] [Q&A]
Hi 
I'm trying to implement new pmd rule where I need to check the very first statement after catch block should be log.error("msg");  if it's not there it should give violation something like this-

```java
try {
    // Block of code to try
}
catch(Exception e) {
    // this line must having logger.error(e)---------> specifically focused on .error only if it's not there then generate violation
}
```

I was able to write this logic using XPath but I want to write it using java @oowekyala 
Can you please help me how can I achieve this I'd be really grateful thank you !

**Comment by adangel** [comment]
When writing the rule in Java, you'll need to traverse the AST with java method calls manually. Your rule extends from `AbstractJavaRule`, which is a visitor. You can the override any visit method to start traversing the tree. In your case, you probably want to start visiting with the catch-statement, so you need to override `public Object visit(ASTCatchStatement node, Object data)`.

This method has a first parameter the current node - in that case the catch statement. You can then ask the node for its immediate children (`getChild(int)`). There are many methods, see the (non-deprecated) methods defined in [Node](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/Node.java) - the common interfaces of all nodes (e.g. getNumChildren, getChild, getParent, getFirstParentOfType, getFirstChildOfType, findDescendantsOfType, ...).
Some node types have also some specific methods, e.g. the class `ASTCatchStatement` has  `getExceptionId()` or `getBody()`, which use the methods mentioned before under the hood.

If you look at the following complete java code sample in the designer:

```java
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class Foo {
    private static final Logger logger = LoggerFactory.getLogger(Foo.class);
    public void bar() {
        try {
            buzz();
        } catch (Exception e) {
            logger.error(e);
        }
    }
}
```

you'll see, that the call to "logger.error" manifests itself as a node of the the "Name" - this corresponds to the java class node type `ASTName`:

![grafik](https://user-images.githubusercontent.com/1573684/203751346-39171f51-bdb2-486a-bdeb-2525e3b5d867.png)

So, you could manually call `getChild(0).getChild(0)....` until you get the node you want to examine. However, using e.g. `node.findDescendantsOfType(...)` is much easier - it collects nodes of a specific type into a list. For this case: `List<ASTName> names = node.findDescendantsOfType(ASTName.class)`. Then you can check each name, whether it is actually a "logger.error" call (`hasImageEqualTo`).
And then further check the arguments of the method call.

A complete rule class could look like this:

```java
package net.sourceforge.pmd.examples.java.rules;

import java.util.List;

import net.sourceforge.pmd.lang.java.ast.ASTArgumentList;
import net.sourceforge.pmd.lang.java.ast.ASTCatchStatement;
import net.sourceforge.pmd.lang.java.ast.ASTName;
import net.sourceforge.pmd.lang.java.ast.ASTPrimaryExpression;
import net.sourceforge.pmd.lang.java.ast.ASTVariableDeclaratorId;
import net.sourceforge.pmd.lang.java.rule.AbstractJavaRule;

public class MyRule extends AbstractJavaRule {

    @Override
    public Object visit(ASTCatchStatement node, Object data) {
        boolean found = false;
        ASTVariableDeclaratorId exceptionId = node.getExceptionId();
        List<ASTName> names = node.getBody().findDescendantsOfType(ASTName.class);
        for (ASTName name : names) {
            if (name.hasImageEqualTo("logger.error") && name.getFirstParentOfType(ASTPrimaryExpression.class) != null) {
                ASTArgumentList argumentList = name.getFirstParentOfType(ASTPrimaryExpression.class).getFirstDescendantOfType(ASTArgumentList.class);
                if (argumentList != null && argumentList.size() > 0) {
                    List<ASTName> refs = argumentList.findDescendantsOfType(ASTName.class);
                    for (ASTName ref : refs) {
                        if (ref.hasImageEqualTo(exceptionId.getName())) {
                            found = true;
                        }
                    }
                }
            }
        }
        if (!found) {
            asCtx(data).addViolation(node);
        }
        return super.visit(node, data);
    }
}
```

**Comment by Prapti08** [comment]
Thanks a lot for this detailed explanation, can you please tell me what was your thought process behind writing this code,  I don't know many of the methods and how to use them to write some complex rules in java. Any resources or topics I need to study for this?

**Reply by adangel** [reply]
Since PMD is an open source project, all the code is available. When you want to write a custom rule, it might make sense to look out for a similar rule that already exists and study its code.
